src/main/resources/static/data-mapping.js

```diff
@@ -16,122 +23,143 @@ $( document ).on( "mousemove", function( event ) {
+            map_data = data;
+        });
     });
 
+    //for testing: mapping data
     $('#auto-map-btn').click(function(){
-        d3.tsv('test-mapping.tsv', function(mapping) {
-            uploaded_mapping = mapping;
-            for(var i=0; i<mapping.length; i++) {
-                var inst = mapping[i];
-                var usr_name = inst['USR'];
-                var oc_name = inst['OC'];
-                clearManualItems(usr_name);
-                for(var j=0; j<ocitems.length; j++) {
-                    var item = ocitems[j];
-                    if(item.ocname == oc_name) {
-                        item.connected = true;
-                        item.usrname = usr_name;
-                        break;
-                    }
-                }//for each mapping-pair in oc-items
-            }//for each mapping instance
-            updateMatchingItems();
-            var is_tree_view = $('#tree-toggle').prop('checked');
-            if(!is_tree_view) {
-                d3.selectAll('.mitem').each(function(d,i){
-                    if(d.connected) {
-                        var m = 5;
-                        d3.selectAll('.griditem').each(function(gridd, gridi) {
-                            if(gridd.name == d.usrname) {
-                                gridd.x = +d.x - 2*m - d.w;
-                                gridd.y = +d.y;
-                                gridd.mapped = true;
-                                gridd.matchingItem = d;
+        clearMapping();
+        for(var i=0; i<map_data.length; i++) {
+            var pair = map_data[i];
+            var ocd = findOCitem(pair['Study_Site'], pair['EventName'], pair['CRFname'], pair['CRFversion'], pair['ocItemName']);
+            var usritem_obj = findUsrItem(pair['usrItemName']);
+            // console.log(ocd);
+
+            if(ocd !== null && usritem_obj !== null) {
+                var ocname = ocd.name;
+                var ocCRFv = ocd.parent.name;
+                var ocCRF = ocd.parent.parent.name;
+                var ocEventName = ocd.parent.parent.parent.name;
+                var ocStudy = ocd.parent.parent.parent.parent.name;
+                var path = ocStudy+"\t"+ocEventName+"\t"+ocCRF+"\t"+ocCRFv+"\t"+ocname;
+                var d = usritem_obj.datum();
+                if(mapped_ocitems.indexOf(path) == -1) {
+                    mapped_ocitems.push(path);
+                    d.mapped = true;
+                    d.ocItemName = ocname;
+                    d.ocCRFv = ocCRFv;
+                    d.ocCRF = ocCRF;
+                    d.ocEventName = ocEventName;
+                    d.ocStudy = ocStudy;
+                    d.ocItemData = ocd;
+                    d.ocPath = path;
+                }
+                positionUsrList(usr_item_data, 800);
+            }
+        }
+
+        function findOCitem(study, event, crf, crfv, itemname) {
+            for(var _study in oc_data.children) {
```

Comment:
Indeed there are drawbacks of using the "in" clause, however,because any undefined element in an array wouldn't be used to construct visual element (as they are undefined), the iteration ignoring the corresponding indices is fine in our code. As for the array prototype change, in the unlikely event where we did make such change, it wouldn't affect the running of the code as we explicitly check the content of each element (study, event, crf, etc.), as I have argued previously. Hope this is clear.